### PR TITLE
Add NoNameShadowing rule

### LINKED ISSUE #240: Rule: Shadowed names
The Kotlin compiler already prints a warning for each shadowed name. Should Detekt also detect these and report them?

Example:
```
override fun bindItemView(position: Int, view: View, list: List<Item>) {        
    setOnClickListener { view -> handleClick(view) }
}
```
The view in the `setOnClickListener` lambda is shadowing the fun parameter.

### PR DESCRIPTION
**Pull request by t-kameyama**
Fixes #240

### CODE REVIEW

File: detekt-core/src/main/resources/default-detekt-config.yml [hunk 1]
```diff
@@ -423,6 +423,8 @@ naming:
   MemberNameEqualsClassName:
     active: true
     ignoreOverridden: true
+  NoNameShadowing:
+    active: false
```
Comment (chao2zhang) [inline detekt-core/src/main/resources/default-detekt-config.yml:427]: +1 for putting this under naming. Thank you!

File: detekt-rules-naming/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowing.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    override val issue = Issue(
+        javaClass.simpleName,
+        Severity.Defect,
+        "Disallows shadowing variable declarations.",
+        Debt.FIVE_MINS
+    )
+
+    override fun visitProperty(property: KtProperty) {
+        super.visitProperty(property)
+        report(property)
```
Comment (chao2zhang) [inline detekt-rules-naming/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowing.kt:48]: Today I learned: `KtProperty` and `DestructuringDeclarationEntry` are the sealed types of `KtVariableDeclaration`

**Review by chao2zhang — APPROVED** [review]

File: detekt-rules-naming/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowing.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import io.gitlab.arturbosch.detekt.api.Severity
+import org.jetbrains.kotlin.diagnostics.Errors
+import org.jetbrains.kotlin.psi.KtDestructuringDeclarationEntry
+import org.jetbrains.kotlin.psi.KtNamedDeclaration
+import org.jetbrains.kotlin.psi.KtParameter
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.resolve.BindingContext
+
+/**
+ * Disallows shadowing variable declarations.
```
Comment (BraisGabin) [inline detekt-rules-naming/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowing.kt:18]: We should elaborate a bit here. I don't know if it's necessary to explain what variable shadowing means. But at least say why it's bad.

File: detekt-rules-naming/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowingSpec.kt [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            val code = """
+                fun test(k: Int) {
+                    listOf(1).map { k ->
+                    }
+                }
+            """
+            val findings = subject.compileAndLintWithContext(env, code)
+            assertThat(findings).hasSize(1)
+            assertThat(findings[0]).hasMessage("Name shadowed: k")
+        }
```
Comment (BraisGabin) [inline detekt-rules-naming/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowingSpec.kt:49]: Does this implementation check the double `it` when you nest two lambdas?

File: detekt-rules-naming/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowing.kt (hunk 3 above)
Comment (t-kameyama) [reply detekt-rules-naming/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowing.kt:18]: bd359d0bac8d0b43e1a1b41806df4ac62e65d1b2

File: detekt-rules-naming/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowingSpec.kt (hunk 4 above)
Comment (t-kameyama) [reply detekt-rules-naming/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowingSpec.kt:49]: 7a794a1ceddb7dcaac1e0efeceeb922a3eea2a5b

File: detekt-rules-naming/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowingSpec.kt [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+                        listOf(2).forEach { it ->
+                        }
+                    }
+                }
+            """
+            val findings = subject.compileAndLintWithContext(env, code)
+            assertThat(findings).hasSize(1)
+            assertThat(findings[0]).hasMessage("Name shadowed: it")
+        }
+
```
Comment (BraisGabin) [inline detekt-rules-naming/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowingSpec.kt:64]:
```suggestion

        it("report shadowing nested lambda implicit 'it' parameter") {
            val code = """
                fun test() {
                    listOf(1).forEach {
                        listOf(2).forEach {
                        }
                    }
                }
            """
            val findings = subject.compileAndLintWithContext(env, code)
            assertThat(findings).hasSize(1)
            assertThat(findings[0]).hasMessage("Name shadowed: it")
        }

```

Do this have sense?

**Review by BraisGabin — APPROVED** [review]

File: detekt-rules-naming/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowingSpec.kt (hunk 5 above)
Comment (t-kameyama) [reply detekt-rules-naming/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowingSpec.kt:64]: Fixed in fec06b6 and 085ab96.

File: detekt-rules-naming/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowing.kt [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        super.visitDestructuringDeclarationEntry(multiDeclarationEntry)
+        report(multiDeclarationEntry)
+    }
+
+    override fun visitParameter(parameter: KtParameter) {
+        super.visitParameter(parameter)
+        report(parameter)
+    }
+
+    private fun report(declaration: KtNamedDeclaration) {
```
Comment (cortinico) [inline detekt-rules-naming/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/naming/NoNameShadowing.kt:72]: Please rename this function. `report` is a method of the superclass and is doing the actual issue reporting. Here you're instead checking for diagnostics and eventually reporting.

**Review by cortinico — APPROVED** [review]

**Review by BraisGabin — APPROVED** [review]

**Pull request merged**